### PR TITLE
feat: support session persistence with ES

### LINKED ISSUE #23760: Implement Session Entity to store session information

### PR DESCRIPTION
**Pull request by maryarm**
## Description

In the current implementation, the Operate and Tasklist components use ES to support sharing sessions for persistent sessions. This is to support a pod shutting down and not forcing the user to login again. We need to be able to consider this within the Camunda Authentication module such that if a user has an existing session then this is loaded into the security context.

## Related issues

closes #23760 #23761 #23762 #23763

### CODE REVIEW

**Comment by maryarm** [comment]
Hey @camunda/hto-backend, I've added you as the reviewer because this task is to move the existing mechanism for web session storage to new arch. Please take a look and also feel free to add relevant interviewers.

File: authentication/src/main/java/io/camunda/authentication/handler/session/SessionDocumentMapper.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      if (attributesObject instanceof Map) {
+        final Map<String, Object> attributes = (Map<String, Object>) document.get(ATTRIBUTES);
+        attributes
+            .keySet()
+            .forEach(
+                name -> {
+                  if (attributes.get(name) instanceof String) {
+                    session.setAttribute(
+                        name,
+                        deserialize(Base64.getDecoder().decode((String) attributes.get(name))));
```
Comment (nathansandi) [inline authentication/src/main/java/io/camunda/authentication/handler/session/SessionDocumentMapper.java:105]: 🔧  Should we add a try/catch in order to catch possible deserialisation errors over here and log that?

File: search/search-client-elasticsearch/src/main/java/io/camunda/search/es/clients/ElasticsearchSessionDocumentClient.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  private final String indexName = String.format("%s-%s-%s_", "camunda", "web-session", "1.1.0");
+
+  public ElasticsearchSessionDocumentClient(final RetryElasticsearchClient client) {
+    this.client = client;
+  }
+
+  public void setup() {
+    final Map<String, Object> indexDescription =
+        ElasticsearchJSONUtil.readJSONFileToMap("/camunda-web-session.json");
+    client.createIndex(
```
Comment (nathansandi) [inline search/search-client-elasticsearch/src/main/java/io/camunda/search/es/clients/ElasticsearchSessionDocumentClient.java:37]: 🔧  Should we add try/catch in order to log possible errors over the **_ES operations_** (in general) ?

File: authentication/src/main/java/io/camunda/authentication/handler/session/NoOpSessionDocumentStorageClient.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class NoOpSessionDocumentStorageClient implements SessionDocumentStorageClient {
+  private final Map<String, Map<String, Object>> sessionDocuments =
+      Collections.synchronizedMap(new HashMap<>());
+
+  @Override
+  public void consumeSessions(final Consumer<Map<String, Object>> sessionConsumer) {
```
Comment (nathansandi) [inline authentication/src/main/java/io/camunda/authentication/handler/session/NoOpSessionDocumentStorageClient.java:21]: 💯 Nice

File: .idea/codeStyles/Project.xml [hunk 4]
```diff
@@ -454,7 +454,6 @@
       </indentOptions>
     </codeStyleSettings>
     <codeStyleSettings language="Python">
-      <option name="KEEP_BLANK_LINES_IN_CODE" value="1" />
```
Comment (nathansandi) [inline .idea/codeStyles/Project.xml:457]: ❓ Is this change will be bring on this PR?

**Review by nathansandi — APPROVED** [review]
Overall, looks good to me
I left some suggestions and questions related to the implementation, and potential improvements.

File: dist/src/main/java/io/camunda/application/commons/search/ElasticSearchSearchClientConfiguration.java [hunk 5]
```diff
@@ -31,4 +37,30 @@ public ElasticsearchSearchClient elasticsearchSearchClient(
     final var elasticsearch = connector.createClient();
     return new ElasticsearchSearchClient(elasticsearch);
   }
+
+  @Bean(destroyMethod = "close")
+  public RestHighLevelClient restHighLevelClient(final SearchClientProperties configuration) {
```
Comment (houssain-barouni) [inline dist/src/main/java/io/camunda/application/commons/search/ElasticSearchSearchClientConfiguration.java:42]: we are actually trying to get rid of ElasticsearchClient 7 dependency (which brings the `RestHighLevelClient`), is it possible to use `co.elastic.clients.elasticsearch.ElasticsearchClient`?

**Review by houssain-barouni — CHANGES_REQUESTED** [review]
❌ Including the `elasticsearch-rest-high-level-client` dependency in the search module is not an option. 
Ideally, `SessionDocumentStorageClient` should have an implementation that is abstracted from ES/OS, though I understand this may require additional effort. 
That said, we should avoid reusing `RetryElasticsearchClient` and instead use `co.elastic.clients.elasticsearch.ElasticsearchClient` from `elasticseach-java` dependency.

File: search/search-client-elasticsearch/src/main/resources/camunda-web-session.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (Zelldon) [inline search/search-client-elasticsearch/src/main/resources/camunda-web-session.json:1]: ❓ Q: Shouldn't this be part of the webapps-schema module ?

https://github.com/camunda/camunda/tree/main/webapps-schema/src/main/resources/schema/elasticsearch/create/index

**Review by houssain-barouni — CHANGES_REQUESTED** [review]
Hi @maryarm 
I am not fully convinced that having a separate Elasticsearch client within the search module solely for session storage is the best approach. I see two alternative options:
1. Manage the session index like any other index, with index creation handled by the Camunda exporter or through a shared module with the exporter for schema management.
2. Treat this index as an internal asset of the authentication module. So you create, inside the authentication module, an ES client that allows you to create the index, and read, write documents in it

@romansmirnov I would like to have your opinion about these kind of indices that need to be managed and fed outside the exporter?

**Comment by maryarm** [comment]
> Hi @maryarm I am not fully convinced that having a separate Elasticsearch client within the search module solely for session storage is the best approach. I see two alternative options:
> 
> 1. Manage the session index like any other index, with index creation handled by the Camunda exporter or through a shared module with the exporter for schema management.
> 2. Treat this index as an internal asset of the authentication module. So you create, inside the authentication module, an ES client that allows you to create the index, and read, write documents in it
> 
> @romansmirnov I would like to have your opinion about these kind of indices that need to be managed and fed outside the exporter?

The first option sounds like a good idea to me (move all schema creation stuff to a shared module), but I find it a big bigger than scope of this PR. 
For the second approach should then we also create an OpensearchClient within authentication module? I mean in my design we have a simple in-memory implementation in authentication module but the other implementation comes in search module based on the configured storage. Maybe the search is not exactly the match but in existing modules I can't find any other relevant module.
that would be nice to hear your ideas here @romansmirnov.

File: search/search-client-elasticsearch/src/main/java/io/camunda/search/es/clients/ElasticsearchSessionDocumentClient.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ElasticsearchSessionDocumentClient implements SessionDocumentStorageClient {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(ElasticsearchSessionDocumentClient.class);
+  private final RetryElasticsearchClient client;
+  private final String indexName = String.format("%s-%s-%s_", "camunda", "web-session", "1.1.0");
```
Comment (Zelldon) [inline search/search-client-elasticsearch/src/main/java/io/camunda/search/es/clients/ElasticsearchSessionDocumentClient.java:24]: ❌ We must support with 8.7 one global prefix https://github.com/camunda/camunda/issues/23807

**Comment by Zelldon** [comment]
> 1. Manage the session index like any other index, with index creation handled by the Camunda exporter or through a shared module with the exporter for schema management.

I think it makes sense to use the Camunda Exporter to do the schema management (creation and validation), I think it is good to have this in one place only. So we don't have this kindof logic everywhere distributed.

**Comment by romansmirnov** [comment]
> I think it makes sense to use the Camunda Exporter to do the schema management (creation and validation), I think it is good to have this in one place only. So we don't have this kindof logic everywhere distributed.

To avoid doing the same multiple times and to avoid having the logic spread across multiple modules, I agree with @Zelldon and @houssain-barouni to have the index creation as part of the schema management in the camunda exporter.

File: authentication/src/main/java/io/camunda/authentication/config/SessionRepositoryConfig.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import jakarta.servlet.http.HttpServletRequest;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+import org.springframework.session.SessionRepository;
+import org.springframework.session.config.annotation.web.http.EnableSpringHttpSession;
+
+@Configuration
```
Comment (romansmirnov) [inline authentication/src/main/java/io/camunda/authentication/config/SessionRepositoryConfig.java:25]: All Spring configurations must be centralized in the `dist` module.

File: dist/src/main/java/io/camunda/application/commons/search/SearchClientDatabaseConfiguration.java [hunk 9]
```diff
@@ -38,11 +46,40 @@ public class SearchClientDatabaseConfiguration {
...
+
+  @Bean
+  @ConditionalOnProperty(
+      prefix = "camunda.database",
+      name = "type",
+      havingValue = "elasticsearch",
+      matchIfMissing = true)
+  @ConditionalOnSessionPersistence
+  @Primary
+  public ElasticsearchSessionDocumentClient elasticsearchSessionDocumentClient(
```
Comment (romansmirnov) [inline dist/src/main/java/io/camunda/application/commons/search/SearchClientDatabaseConfiguration.java:74]: Why do we need this separation? Why do we need a `RetryElasticsearchClient`?

File: authentication/src/main/java/io/camunda/authentication/config/SessionRepositoryConfig.java (hunk 8 above)
Comment (maryarm) [reply authentication/src/main/java/io/camunda/authentication/config/SessionRepositoryConfig.java:25]: If it's the standard then Ok, but as we have WebSecurityConfig in authentication module for a long time, should it be moved either?

File: search/search-client-elasticsearch/src/main/java/io/camunda/search/es/clients/ElasticsearchSessionDocumentClient.java (hunk 7 above)
Comment (maryarm) [reply search/search-client-elasticsearch/src/main/java/io/camunda/search/es/clients/ElasticsearchSessionDocumentClient.java]: Good point, I have renamed the index to start with identity as I didn't want to add a new component.

**Pull request closed without merging**